Update to 2.0.0

WillieCubed opened the pull request:
This is a running PR for the changes needed for the 2.0.0 release.
 - [ ] ~Create maintenance v1.1.1 release to migrate users to v2.0.0~ 
 - [x] Migrate to Android Architecture Components: #2 
 - [x] Migrate to Cloud Firestore: #1
 - [ ] Redesign app: #3 
 - [ ] Final refactor #7

WillieCubed committed:
Revise class documentation code style
 - Also fix line endings

WillieCubed committed:
Downgrade version to match FirebaseUI's library versions
 - Fix build errors

WillieCubed committed:
Create deserializers to map DataSnapshots to Countdowns

WillieCubed committed:
Create ViewModels to handle Activity/Fragment states

WillieCubed committed:
Update build.gradle files

WillieCubed committed:
Update CountdownListActivity to use SelectedCountdownViewModel

WillieCubed committed:
Deprecate DetailSheetController and CountdownPersistentDetailFragment

WillieCubed committed:
Fix onSaveInstanceState error

WillieCubed committed:
Create ModalCountdownButtonSheet to handle deep-linked Countdowns
 - Update CountdownDeepLinkActivity

WillieCubed committed:
Update CountdownDetailFragment to be persistent bottom sheet

WillieCubed committed:
Replace DetailSheetController with CountdownDetailClickHandler

WillieCubed committed:
Merge branch 'finish-refactor' into 2.0.0-dev

WillieCubed committed:
Merge branch 'master' into 2.0.0-dev

WillieCubed committed:
Deprecate OldDatabase that uses Firebase RTDB

WillieCubed committed:
Merge remote-tracking branch 'origin/master' into cloud-firestore

WillieCubed committed:
Create UserRepository as source for user data

WillieCubed committed:
Deprecate Users

WillieCubed committed:
Update data models
 - Update Countdown to support dedicated UID
 - Create User to decouple from FirebaseUser

WillieCubed committed:
Clean up user management
 - Create AuthFlowManager to handle UI-facing auth management
 - Create UserManager to replace gross utility Users class

WillieCubed committed:
Deprecate old database classes

WillieCubed committed:
Extract strings from AuthFlowManager

WillieCubed committed:
Add data clear methods to UserRepository

WillieCubed committed:
Fix deprecated class usage in SettingsActivity
 - Remove reliability on FirebaseUser
 - Make Indexer return the a task
 - Use UserRepository to clear user data

WillieCubed committed:
Create CountdownDocumentDeserializer and CountdownListDocumentDeserializer

WillieCubed committed:
Create CountdownRepository to replace CountdownManager
 - Migrate most CountdownManager functions
 - Handle countdown creation, updating, fetching, deleting

WillieCubed committed:
Update CountdownDetailFragment to use better ViewModel initialization

WillieCubed committed:
Fix CountdownListWidgetService

WillieCubed committed:
Update CountdownRepository
 - Improve uploadCountdown semantics

WillieCubed committed:
Deprecate Indexer#indexCountdown(Countdown, String, FirebsaseUser)
 - update CountdownCreationActivity

WillieCubed committed:
Update User model to include fcmToken attribute

WillieCubed committed:
Update UserRepository to support adding FCM tokens to users
 - also update CountdownInstanceIdService

TheCraftKid commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Countdowns v2.0.0](https://trello.com/c/y6jj6Knq/1-countdowns-v200)

TheCraftKid commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Version 2.0.0](https://trello.com/c/oRYqmQY1/2-version-200)

WillieCubed committed:
Create QuerySource to house queries for common database resources
 - Update UserRepository to use QuerySource
 - Update CountdownRepository to use QuerySource

WillieCubed committed:
Replace FirebaseQueryLiveData with FirestoreQueryLiveData
 - Support Cloud Firestore

WillieCubed committed:
Update CountdownRecyclerAdapter for Cloud Firestore

WillieCubed committed:
Update countdown lists to use new CountdownRecyclerAdapter constructor

WillieCubed committed:
Fix nullability issue

WillieCubed committed:
Fix CountdownListDeserializer
 - Fix silly reference to RTDB

WillieCubed committed:
Update StartActivity to use AuthFlowManager

WillieCubed committed:
Deprecate more classes

WillieCubed committed:
Fix deprecated usages

WillieCubed committed:
Update ViewModels
 - Remove unused CountdownListViewModel

WillieCubed committed:
Update UserPrivileges to use Firestore

WillieCubed committed:
Dump buildable code

WillieCubed committed:
Update git submodule

WillieCubed committed:
Update git submodule

WillieCubed committed:
Make Travis happy

WillieCubed committed:
Move to Gradle dependency for CraftEssentials

WillieCubed committed:
Update dependencies

WillieCubed commented:
I'm going to rewrite the app from scratch.
